res/controllers/PIONEER-DDJ-ERGO-scripts.js

```diff
@@ -0,0 +1,441 @@
+//this also deals with LEDs, turning on the FX slot button currently active and the ON button
+DDJERGO.effectsFunction = function(channel, control, value, status) {
+    const deck = status-0x94;
+    const effect = selectedEffects[deck];
+    //on button pressed
+    if (control === 0x43 && value === 0x7F) {
+        const toggleState = 1 - engine.getParameter(effectsStrings[deck][effect], "enabled");
+        engine.setParameter(effectsStrings[deck][effect], "enabled", toggleState);
+        //toggle ON button and selected effect button
+        midi.sendShortMsg(status, 0x43, 0x7F*toggleState);
+        midi.sendShortMsg(status, 0x47+effect, 0x7F*toggleState);
+    }
+    //fx number button pressed
+    else if (control >= 0x47 && value === 0x7F) {
+        selectedEffects[deck] = control - 0x47;
+    }
+};
+
+//this modifies the selected effects parameters based on the MSBs of the FX knobs
+DDJERGO.effectsParams = function(channel, control, value, status) {
+    const deck = status-0xB4;
+    const effect = selectedEffects[deck];
+    //if called with MSB
+    if (control < 0x20) {
+        //if fx selector turned enough
+        if (control === 0x00 && previousEffectKnobPosition !== value/5) {
+            //switch to next or previous fx in list
+            if (previousEffectKnobPosition > value/5) {
+                engine.setValue(effectsStrings[deck][effect], "effect_selector", -1);
+            } else {
+                engine.setValue(effectsStrings[deck][effect], "effect_selector", 1);
+            }
+            previousEffectKnobPosition = value/5;
+        }
+        //if fx knob 1 turned
+        if (control === 0x02) {
+            engine.setValue(effectsStrings[deck][effect], "meta", value/127);
+        }
+    }
+};
+
+//this plays the samples on the samplers with a click on the samplers buttons and stores its value for the sample vol knob to manipulate it
+DDJERGO.samplers = function(channel, control, value, status, group) {
+    const deck = status-0x90;
+    if (value === 0x7F) {
+        if (engine.getParameter(group, "play") === 1) {
+            engine.setParameter(group, "cue_default", 1);
+        } else {
+            engine.setParameter(group, "play", 1);
+        }
+        //store sample numbers
+        selectedSamples[0] = (control - 0x3C)/2 + (deck)*4;
+        selectedSamples[1] = (control - 0x3C)/2 + (deck)*4;
+    }
+};
+
+//this modifies the selected sample volume with sample vol knob rotations
+DDJERGO.samplerVolume = function(channel, control, value, status) {
+    const deck = status-0xB0;
+    let newValue;
+    if (value-64 > 0) { newValue = value-128; } else { newValue = value; }
+    if (newValue > 0) {
+        script.triggerControl(samplerStrings[selectedSamples[deck]], "pregain_up");
+    }
+    if (newValue < 0) {
+        script.triggerControl(samplerStrings[selectedSamples[deck]], "pregain_down");
+    }
+};
```

Comment:
So this is an rotary encoder (endless), not a pot?

Reply:
yes